monai/data/image_reader.py

```diff
@@ -0,0 +1,348 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+from monai.data.utils import correct_nifti_header_if_necessary
+from monai.utils import ensure_tuple, optional_import
+
+itk, _ = optional_import("itk", allow_namespace_pkg=True)
+nib, _ = optional_import("nibabel")
+
+
+class ImageReader(ABC):
+    """Abstract class to define interface APIs to load image files.
+    users need to call `read` to load image and then use `get_data`
+    to get the image data and properties from meta data.
+
+    Args:
+        img: image to initialize the reader, this is for the usage that the image data
+            is already in memory and no need to read from file again, default is None.
+
+    """
+
+    def __init__(self, img: Optional[Any] = None) -> None:
+        self.img = img
+        self._suffixes: Optional[Sequence[str]] = None
+
+    def get_suffixes(self) -> Sequence[str]:
+        """
+        Get the supported image file suffixes of current reader.
+        Default is None, support all kinds of image format.
+
+        """
+        return self._suffixes
+
+    def verify_suffix(self, filename: str) -> bool:
+        """
+        Verify whether the specified file or files match supported suffixes.
+        If supported suffixes is None, skip the verification and return True.
+
+        Args:
+            filename: file name or a list of file names to read.
+                if a list of files, verify all the subffixes.
+        """
+
+        if self._suffixes is None:
+            return True
+
+        supported_format: bool = True
+        filenames: Sequence[str] = ensure_tuple(filename)
+        for name in filenames:
+            suffixes: Sequence[str] = name.split(".")
+            passed: bool = False
+            for i in range(len(suffixes) - 1):
+                if ".".join(suffixes[-(i + 2) : -1]) in self._suffixes:
+                    passed = True
+                    break
+            if not passed:
+                supported_format = False
+                break
+
+        return supported_format
+
+    def uncache(self) -> None:
```

Comment:
it's not very clear why all readers need an 'uncache' method, could we remove this method for now?

Reply:
Hi @wyli ,

We need to call `uncache()` for Nibabel reader, should we add some `additional_operation` for all the readers and then call it in `LoadImage`?

Thanks.

Reply:
i think we should keep the interface simple and clean for now

Reply:
OK, removed in the latest commit according to:
https://nipy.org/nibabel/images_and_memory.html#use-the-array-proxy-instead-of-get-fdata

Thanks.